File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(!root){
            return true;
        }
        int leftDepth = depth(root->left);
        int rightDepth = depth(root->right);
        if(abs(leftDepth - rightDepth) > 1){
            return false;
        }
        return isBalanced(root->left) && isBalanced(root->right);
    }

    int depth(TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        return max(depth(root->left), depth(root->right)) + 1;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        left = 0 
        right = len(heights) - 1
        maxArea = 0

        def calcArea(left, right):
            area = (right-left) * min(heights[right], heights[left])

            return area

        
        while left < right:
            curr = calcArea(left, right)
            
            maxArea = max(maxArea, curr)

            if heights[left] < heights[right]:
                left += 1
            else:
                right -= 1
            
        return maxArea